Fix an oversight regarding building unlocking

penev92 opened the pull request:
Fix an oversight regarding building unlocking

penev92 committed:
Fix an oversight regarding building unlocking

pchote commented on OpenRA.Mods.Common/Traits/Transforms.cs:
```diff
@@ -127,11 +128,21 @@ public void DeployTransform(bool queued)
+			{
+				// We locked it at the start of this method, so we need to unlock it now.
+				if (building != null)
+					building.Unlock();
```
This leaves the actor unlocked for a tick, which opens us up to a bunch of edge cases (e.g. selling / capturing the conyard immediately before it turns into an MCV).

You don't need to unlock the actor because it gets discarded when the transform succeeds.  I would set this up like:

```
queue(turn)
queue(lock)
queue(play unmake animation)
queue(check footprint, and transform if allowed)
queue(make animation)
queue(unlock)
```

If the transform succeeds then the activities after that won't run.  If the transform fails then it will restore itself to a valid state.

penev92 commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/Transforms.cs:
```diff
@@ -127,11 +132,23 @@ public void DeployTransform(bool queued)
+				}
+			});
+
 			if (makeAnimation != null)
```
As mentioned in the earlier discussion you also want to check this *before* running the make animation to filter out the cases that will always fail (e.g. trying to deploy next to invalid cells).

penev92 replied:
That only happens for unmake animations, and only if you unmake from a building to an actor that is bigger than that building. Make animations aren't played by  `Transforms`, just unmake animations.

pchote replied:
We should have a lint test that ensures that assumption remains true (but not necessary for this pr).

pchote commented on OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs:
```diff
@@ -39,12 +39,18 @@ public WithMakeAnimation(ActorInitializer init, WithMakeAnimationInfo info)
+				Forward(self, building);
+		}
+
+		// TODO: All the Building stuff that uses make animations should go through this class,
```
Through this method?

pchote commented on OpenRA.Mods.Common/Traits/Transforms.cs:
```diff
@@ -96,7 +100,7 @@ public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool qu
 		{
 			var building = self.TraitOrDefault<Building>();
-			if (!CanDeploy() || (building != null && !building.Lock()))
+			if (!CanDeploy(true) || (building != null && !building.Lock()))
```
This check should only be done if `queued == false`.

pchote replied:
Actually, this requires a bigger change to handle the building locking... so maybe not a good idea for the prep branch.  This trait needs a total rewrite in the near future.

obrakmann commented:
Seems to work. :+1: 

I'm feeling iffy about taking this into stable, though. This code has regressed in a million subtle ways before, and I fear we'll just be trading one bug for another.

pchote commented:
I have mentioned in IRC and the earlier discussion here that there is a simpler short-term fix that would be more appropriate to merge to prep: keep everything else the same, but include a check for "can deploy right now" before starting the unmake animation / actual transform.

Phrohdoh commented:
Adding `[PR: Changes requested]` at least for https://github.com/OpenRA/OpenRA/pull/9918#discussion_r44338007, I'm not sure if the above comment is just a statement or asking for the mentioned route to be taken.

Mailaender commented:
I don't know if this was completely superseded by https://github.com/OpenRA/OpenRA/issues/9987 or is still improving things after a rebase.

penev92 commented:
The main idea behind this was done with the other PR. I still plan to salvage some things from this though.

penev92 commented:
Although the issue this PR was originally intended to fix was already fixed by @pchote, I decided to repurpose it and keep the little fix for something I stumbled on while debugging the original issues.
Removed the second commit, which is now irrelevant and updated the PR's name and description.

abcdefg30 commented:
Well, that doesn't change anything*, but won't hurt either. :+1: 

*`Unlock()` just does `Locked = false;`

penev92 commented:
That's the idea. This is accessing something that it shouldn't.

obrakmann commented:
:+1:

Phrohdoh commented:
If this should not be accessing `Locked` it should not be public (even though that wouldn't fix this case).

penev92 commented:
Chances are its setter should indeed be private to avoid such mistakes in the future.
